packages/test/stochastic-test-utils/src/markovChain.ts

```diff
@@ -0,0 +1,283 @@
+                    } else {
+                        this.chain[prevWord][word] = 1;
+                    }
+                    prevWord = word;
+                }
+            }
+
+            if (sentence.length === 1) {
+                const word = sentence[0];
+                if (this.chain[word][MarkovChain.MARKOV_SENTENCE_END_KEY] !== undefined) {
+                    this.chain[word][MarkovChain.MARKOV_SENTENCE_END_KEY] += 1;
+                } else {
+                    this.chain[word][MarkovChain.MARKOV_SENTENCE_END_KEY] = 1;
+                }
+            }
+        });
+    }
+
+    /**
+     * Runtime per word added to the generated sentence: O(totalNumberOfWordChoices) + O(wordLength)
+     * @returns A sentence generated using the given class instances markov chain.
+     */
+    public generateSentence(maxLength: number) {
+        const markovChain = this.chain;
+        if (Object.keys(markovChain).length === 0 || markovChain[MarkovChain.MARKOV_SENTENCE_BEGIN_KEY] === undefined) {
+            return "";
+        }
+
+        let sentence = "";
+        const rootWordChoices = markovChain[MarkovChain.MARKOV_SENTENCE_BEGIN_KEY];
+        if (Object.keys(rootWordChoices).length === 0) {
+            throw Error("Provided markov chain because it has no root words");
+        }
+        const rootWord = this.randomlySelectWord(rootWordChoices);
+
+        sentence += rootWord;
+        let currWord = rootWord;
+        let currWordSpacing = MarkovChain.assumeWordLanguageSpacing(currWord);
+        let prevWordSpacing = currWordSpacing;
+        let nextWordChoices = markovChain[currWord];
+        while (sentence.length < maxLength && Object.keys(nextWordChoices).length !== 0) {
+            prevWordSpacing = currWordSpacing;
+            currWord = this.randomlySelectWord(nextWordChoices);
+            if (currWord === MarkovChain.MARKOV_SENTENCE_END_KEY) {
+                break;
+            }
+            currWordSpacing = MarkovChain.assumeWordLanguageSpacing(currWord);
+
+            if (currWordSpacing === "UNKNOWN") {
+                if (prevWordSpacing === "NO_SPACES") {
+                    sentence += `${currWord}`;
+                } else {
+                    sentence += ` ${currWord}`;
+                }
+            } else if (currWordSpacing === "NO_SPACES") {
+                sentence += `${currWord}`;
+            } else if (currWordSpacing === "SPACED") {
+                sentence += ` ${currWord}`;
+            }
+
+            nextWordChoices = markovChain[currWord];
+        }
+
+        return sentence;
+    }
+
+    private randomlySelectWord(wordOccuranceMap: Record<string, number>) {
+        const wordChoices: string[] = [];
+        Object.entries(wordOccuranceMap).forEach(([word, occuranceCount]) => {
+            for (let i = 0; i < occuranceCount; i++) {
+                wordChoices.push(word);
+            }
+        });
```

Comment:
Have a look at @Abe27342's weighted generator to see how to do this without allocating and populating an array on each call (or possibly change your map to a Record<string, Generator<string>> and use the existing weighted generators.

Reply:
I have updated SpaceEfficientGenerator to utilize the weighted generator. I'm curious what the runtime for that generator is but I assume its better than O(n)?

For the case where we have premade generators ( Record<string, Generator> ) My main concern is whether the generator object is serializable. If it is I can make the Performance Markov chain fit use the premade generator.

Reply:
Sampling the subgenerator to use for `createWeightedGenerator(weights)` is currently `O(weights.length)` where all arrays are pre-allocated and no acceptance criteria are used (there aren't smarts with acceptance criteria; it just rejection samples but this could be improved). It would be straightforward to convert it to `O(log(weights.length))`, I just didn't bother.

Generators are functions so not serializable.

I'm hoping to get a chance to look at the PR later in more detail, but if you have any other quick questions I can answer them :)